Implement identity (recovery) request

bisgardo opened the pull request:
Adds the functions with the following UDL signatures:

- `string identity_issuance_request_json(IdentityIssuanceRequestParameters params)`
- `string identity_recovery_request_json(IdentityRecoveryRequestParameters params)`
- `AccountCredentialResult account_credential(AccountCredentialParameters params)`
- `string account_credential_deployment_hash_hex(AccountCredential credential, u64 expiry_unix)`
- `string account_credential_deployment_signed_payload_hex(SignedAccountCredential credential)`

The parameter types mirror the corresponding "input" types from `wallet_library` but use only types supported by UniFFI. The values are translated into these library types via JSON encoding/decoding. This is in contrast to the Java SDK where the value is passed as a JSON encoded string which is then decoded directly into the library input type. Doing it this way ensures that type unsafe conversions happen internally in this library where it's easily tested rather than at the FFI boundary. So it makes the FFI boundary type safe and of course also generates the Swift types that we do need on the SDK side anyway.

The identity request functions return their result as JSON encoded strings because the protocol actually is to just send the object as JSON in a URL parameter. So there's no point in decoding them into structured types - they would just be converted right back to JSON on the SDK side. We do decode the payload in a unit test to verify the format.

The existing functions have been renamed and reordered to match the conventions and usage in the SDK, so this is a breaking change. Their implementations have not changed.

bisgardo committed:
Implement identity issuance/recovery request and credential deployment functions

UDL signatures:

- `string identity_issuance_request_json(IdentityIssuanceRequestParameters params)`
- `string identity_recovery_request_json(IdentityRecoveryRequestParameters params)`
- `AccountCredentialResult account_credential(AccountCredentialParameters params)`
- `string account_credential_deployment_hash_hex(AccountCredential credential, u64 expiry_unix)`
- `string account_credential_deployment_signed_payload_hex(SignedAccountCredential credential)`

The parameter types mirror the corresponding "input" types from `wallet_library` but use only types supported by UniFFI.
The values are translated into these library types via JSON encoding/decoding.
This is in contrast to the Java SDK where the value is passed as a JSON encoded string
which is then decoded directly into the library input type.
Doing it this way ensures that type unsafe conversions happen internally in this library where it's easily tested
rather than at the FFI boundary.
So it makes the FFI boundary type safe and of course also generates the Swift types that we do need on the SDK side anyway.

The identity request functions return their result as JSON encoded strings
because the protocol actually is to just send the object as JSON in a URL parameter.
So there's no point in decoding them into structured types - they would just be converted right back to JSON on the SDK side.
We do decode the payload in a unit test to verify the format.

bisgardo committed:
Rename and reorder functions

bisgardo commented:
Note that the PR is stacked on https://github.com/Concordium/concordium-wallet-crypto-swift/pull/8.

bisgardo commented on tests/lib_test.rs:
```diff
@@ -14,131 +16,155 @@ const COMMITMENT_KEY: &str = "b14cbfe44a02c6b1f78711176d5f437295367aa4f2a8c2551e
     );
 }
 
-#[test]
-fn testnet_cred_id_matches_cred_deployment() {
-    assert_eq!(
-        get_credential_id(SEED.to_string(), TESTNET.to_string(), 0, 0, 1, COMMITMENT_KEY.to_string()).unwrap(),
-        "b317d3fea7de56f8c96f6e72820c5cd502cc0eef8454016ee548913255897c6b52156cc60df965d3efb3f160eff6ced4"
-    );
-}
```
This test was redundant and has therefore beed removed.

bisgardo committed:
Merge branch 'main' into identity-recovery-request

bisgardo committed:
Merge branch 'main' into identity-recovery-request

bisgardo commented:
Ping @eb-concordium @abizjak, I'm currently blocked by this to proceed on the SDK side. If it helps I can confirm that it all works from there.

bisgardo commented:
> Note that the PR is stacked on #8.

It's no longer stacked.

abizjak commented on CHANGELOG.md:
```diff
@@ -12,11 +12,35 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+  The values are translated into these library types via JSON encoding/decoding.
+  This is in contrast to the Java SDK where the value is passed as a JSON encoded string
+  which is then decoded directly into the library input type.
+  Doing it this way ensures that type unsafe conversions happen internally in this library where it's easily tested
+  rather than at the FFI boundary.
```
I don't understand what difference this is pointing out since the FFI boundary is hidden in the Java SDK as well, in the internals of the library.

bisgardo replied:
The difference is that we have types checked statically instead of at runtime. I'll rephrase.

bisgardo replied:
done

abizjak commented on src/lib.rs:
```diff
@@ -137,19 +151,389 @@ pub fn get_verifiable_credential_public_key(
     })
 }
+
+pub type ArIdentity = u32;
+pub type IpIdentity = u32;
+pub type KeyIndex = u8;
```
Why do these definitions need to be duplicated?

bisgardo replied:
They're not exposed by wallet-library. I thought it was better to do a little copy than take on another dependency. But as all the concordium-base libs are always in sync there's no problem with that so I'll change it.

bisgardo replied:
Turns out that because the types in concordium-base wraps the types (in contrast to being type aliases) complicates their use in a way that isn't worth it. Besides, these aliases were a rather arbitrary subset of the used types anyway, so I inlined them instead.

bisgardo replied:
done

abizjak commented on src/lib.rs:
```diff
@@ -137,19 +151,389 @@ pub fn get_verifiable_credential_public_key(
+    pub attribute_randomness_hex: HashMap<String, String>,
+    #[serde(rename = "credentialPublicKeys")]
+    pub credential_public_keys_hex: CredentialPublicKeysHex,
+}
+
+#[derive(Debug, Serialize)]
+pub struct IdentityObject {
+    #[serde(rename = "preIdentityObject")]
+    pub pre_identity_object: PreIdentityObject,
+    #[serde(rename = "attributeList")]
+    pub attribute_list: AttributeList,
+    #[serde(rename = "signature")]
+    pub signature_hex: String,
+}
+
+#[derive(Debug, Serialize)]
+pub struct PreIdentityObject {
+    #[serde(rename = "idCredPub")]
+    pub id_cred_pub_hex: String,
+    #[serde(rename = "ipArData")]
+    pub ip_ar_data: HashMap<ArIdentity, ArData>,
+    #[serde(rename = "choiceArData")]
+    pub choice_ar_data: ChoiceArParameters,
+    #[serde(rename = "idCredSecCommitment")]
+    pub id_cred_sec_commitment_hex: String,
+    #[serde(rename = "prfKeyCommitmentWithIP")]
+    pub prf_key_commitment_with_ip_hex: String,
+    #[serde(rename = "prfKeySharingCoeffCommitments")]
+    pub prf_key_sharing_coeff_commitments_hex: Vec<String>,
+    #[serde(rename = "proofsOfKnowledge")]
+    pub proofs_of_knowledge_hex: String,
+}
+
+#[derive(Debug, Deserialize, Serialize, PartialEq)]
+pub struct ChoiceArParameters {
+    #[serde(rename = "arIdentities")]
+    pub ar_identities: Vec<ArIdentity>,
+    #[serde(rename = "threshold")]
+    pub threshold: u32,
+}
```
Why do these definitions need to be "duplicated". They seem to have the same JSON representation as the real ones, with fewer consistency checks.

It's a nightmare for maintenance to have all of these complex types duplicated here so there really must be a very good reason for doing so.

We have not had that reason in other SDKs, so not sure why it exists here.

bisgardo replied:
UniFFI requires using (structs of) primitive types or types that implement a specific trait; see https://mozilla.github.io/uniffi-rs/udl/custom_types.html. And I didn't want to propose implementing this trait all over the place. I think that would be much harder to use and understand.

UniFFI will generate swift types that are directly usable in the SDK which reduces (manual) duplication. And in the future you could use the project to generate similar types in other languages.

abizjak commented on src/lib.rs:
```diff
@@ -137,19 +151,389 @@ pub fn get_verifiable_credential_public_key(
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct VerifyKeyWithScheme {
+    #[serde(rename = "schemeId")]
+    pub scheme_id: String,
+    #[serde(rename = "verifyKey")]
+    pub key_hex: String,
+}
```
The public key, as it is now, is expected to be deserializable from either a string, defaulting to ed25519 scheme, or from an object with a scheme tag like this.

Since it's unclear what the purpose of this type is, and how it will be used, it's hard to know whether this difference is relevant.

bisgardo replied:
It only exists because the JSON serialization of `AccountCredential` uses it (via `CredentialPublicKeysWithScheme`). Unlike in the input of the request function where it's just a string. But from what you say I understand that you could also provide the tagged variant as input?

bisgardo replied:
That worked, consolidated the types.

abizjak commented on src/lib.rs:
```diff
@@ -137,19 +151,389 @@ pub fn get_verifiable_credential_public_key(
+    credential: AccountCredential,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Randomness {
+    #[serde(rename = "attributesRand")]
+    pub attributes_rand_hex: HashMap<String, String>,
+    #[serde(rename = "credCounterRand")]
+    pub cred_counter_rand_hex: String,
+    #[serde(rename = "idCredSecRand")]
+    pub id_cred_sec_rand_hex: String,
+    #[serde(rename = "maxAccountsRand")]
+    pub max_accounts_rand_hex: String,
+    #[serde(rename = "prfRand")]
+    pub prf_rand_hex: String,
+}
```
This also seems to be duplicating things by replacing types with Strings.

bisgardo replied:
Yes, in order to conform with UniFFI requirements. Remember that these types exist for being used in the other side of FFI. There's nothing other than string that reasonably makes sense across that boundary.

abizjak replied:
Right, but now you've duplicated everything, documetned nothing, and manually writtent this both in Rust and in the interface description file.

There is also no clearly stated relationship between any of these types and the ones defined in base.

If this UniFFI can generate types, why are we manually defining them here?

Also, the way we convert from imprecise types to precise ones in other places, e.g., when using grpc generated types, is to have TryFrom or From instances that establish a precise relationship between types. It also means that we get compilation errors if either the generated ones or  the more precise defined ones change.

So why can't we do that here? As I read this https://mozilla.github.io/uniffi-rs/tutorial/Rust_scaffolding.html that should be the intended way to do it. Or what am I missing?

bisgardo replied:
Discussed offline.

abizjak commented on src/lib.rs:
```diff
@@ -137,19 +151,389 @@ pub fn get_verifiable_credential_public_key(
+        })
+}
+
+/// Duplicate of [`CredentialDeploymentDetails`] because that type only supports construction from JSON.
```
Unclear what you mean here, the type in the wallet library does support both serialization and deserialization.

Even if it did not the correct solution is to fix it, not introduce a new type that duplicates it.

bisgardo replied:
It doesn't support explicit construction which is what I need here.

abizjak replied:
I don't know what you mean by explicit construction. Do you mean that the fields are currently private? 

This is something that is fixable, and should be fixed rather that worked around like this.

bisgardo replied:
This is only used temporarily in the following function. Do you insist?

bisgardo replied:
Will leave it like this for now, it can always be changed.

abizjak commented on src/lib.rs:
```diff
@@ -137,19 +151,389 @@ pub fn get_verifiable_credential_public_key(
+}
+
+/// Duplicate of [`CredentialDeploymentDetails`] because that type only supports construction from JSON.
+/// It's only used internally and not exported.
```
Well, it's `pub`.

bisgardo replied:
will change

bisgardo replied:
done

abizjak commented on src/lib.rs:
```diff
@@ -137,19 +151,389 @@ pub fn get_verifiable_credential_public_key(
+    pub credential: AccountCredential,
+}
+
+pub fn account_credential_deployment_hash_hex(
```
Unclear what this function is doing. Needs documentation.

Also, what is `expiry_unix`? Is it `s` or `ms` or some other unit?

bisgardo replied:
I intend to add more documentation later but I want to get the SDK done first.

bisgardo replied:
It's Unix time, i.e. seconds since Unix epoch: https://en.wikipedia.org/wiki/Unix_time. I can add `_secs` if that makes it clearer?

bisgardo replied:
done

abizjak commented on src/lib.rs:
```diff
@@ -137,19 +151,389 @@ pub fn get_verifiable_credential_public_key(
+            call: "account_credential_deployment_signed_payload_hex(...)".to_string(),
+            msg: format!("{:#}", e),
+        })
+}
```
In general I do not understand the reasoning for duplicating so much stuff here.

The wallet-library https://github.com/Concordium/concordium-base/tree/main/rust-src/wallet_library/src was specifically to be used by SDKs to expose the relevant identity functionality.

It has a number of types and exposed functions to help with this.

I don't think we want two versions of all of this.

bisgardo replied:
As already answered in other comments, the reason is UniFFI. And the reason why I want to use that is that it allows us to generate *a lot* of boilerplate on the Swift side. Not just types but also very error prone scaffolding around the FFI calls. The functions are exposed simply as regular Swift functions, including proper exceptions.

And because UniFFI supports several other languages, this is all reusable.

I'm open to doing things in a simpler way if that is possible. Duplication is definitely not ideal, but as I see it, the alternative is implementing `UniffiCustomTypeConverter` for the existing types which I think would be much more complex. I do not want to do all the things that UniFFI does for me by hand just to avoid duplicating some structs.

abizjak replied:
Ok.

I am fine with using UniFFI since you spent all this time on it now. I don't think the rationale about "other languages" is relevant here though since the supported ones we already actually have and we won't be rewriting things.

But I still don't understand why you are not using UniFFI code generation as I read it should be used. The Rust types you are defining manually.

bisgardo replied:
Discussed offline.

abizjak commented on tests/lib_test.rs:
```diff
@@ -165,8 +191,528 @@ fn testnet_verifiable_credential_signing_key() {
+struct IdentityRecoveryRequestResult {
+    #[serde(rename = "idRecoveryRequest")]
+    value: Versioned<IdentityRecoveryRequestResultValue>,
+}
```
I'm sorry, but this seems like a horrible to maintain way to test.

Commenting out some fields of some types and comparing the rest for equality is both lossy, and hard to maintain as I see it.

A simple way to, without any boilerplate is to verify the requests, or parse them into proper structures for example if all you care about to test is that it is well-formed.

bisgardo replied:
I didn't know the "proper" types when I started, so didn't consider that. Will change.

bisgardo replied:
done

abizjak reviewed (changes requested):
I might be missing something, but I don't understand the need for all the duplication that is done here, compared to other SDKs we already have.

bisgardo committed:
Various cleanup to address review

- Inlined type aliases: They were a somewhat arbitrary set of the used types anyway and weren't used for FFI. There are similar types in concordium-base except those are not aliases which means that UniFFI doesn't support them automatically.
- Reconcile `CredentialPublicKeysHex` with `CredentialPublicKeysWithScheme` into `CredentialPublicKeys` as it turns out that the string (hex) variant is unnecessary because the scheme variant can be used instead of it.
- Unexport internal struct`CredentialDeploymentPayloadHashInput` and rename its field `expiry_unix` to `expiry_unix_secs` to clarify the unit.
- Remove test structs that are just boiled down versions of the actual types. This changes the tests to verify the complete format but no longer any concrete values.
- Changelog: Minor clarification of rationale.

bisgardo committed:
Merge branch 'main' into identity-recovery-request

abizjak commented on src/lib.udl:
```diff
@@ -3,25 +3,185 @@ enum ConcordiumWalletCryptoError {
     "CallFailed"
 };
 
+dictionary IdentityIssuanceRequestParameters {
```
If these types are to be used for code generation, would the normal thing be to have documentation attached to them, so presumably the generated code has that?

bisgardo replied:
Yes, added a ton of docstrings.

abizjak commented on tests/lib_test.rs:
```diff
@@ -165,8 +191,381 @@ fn testnet_verifiable_credential_signing_key() {
+    version: u32,
+    #[serde(rename = "value")]
+    value: T,
+}
```
This exists in concordium-base
```rust

/// [`Versioned<T>`](Versioned) represents `T` as a versioned data-structure.
/// The version is a integer number up to the implementation,
/// which is serialized using variable integer encoding.
/// The caller is responsible for ensuring the data structure `T`
/// is compatible with the version number.
#[derive(Clone, Copy, Debug, Eq, PartialEq, SerdeSerialize, SerdeDeserialize)]
pub struct Versioned<T> {
    #[serde(rename = "v")]
    pub version: Version,
    #[serde(rename = "value")]
    pub value:   T,
}
```

bisgardo replied:
removed

abizjak commented on tests/lib_test.rs:
```diff
@@ -165,8 +191,381 @@ fn testnet_verifiable_credential_signing_key() {
+
+    // The result is almost entirely non-reproducible,
+    // so the test only verifies that the result was computed successfully and that result has the correct JSON format.
+    serde_json::from_str::<IdentityIssuanceRequestResult>(&res_json).unwrap();
```
You could just verify the request. That's a much better test anyhow that checking some equality of fields.

You have all the context necessary to call
```rust
pub fn validate_id_recovery_request<P: Pairing, C: Curve<Scalar = P::ScalarField>>(
```

bisgardo replied:
done

abizjak commented on tests/lib_test.rs:
```diff
@@ -165,8 +191,381 @@ fn testnet_verifiable_credential_signing_key() {
+
+    // The result is almost entirely non-reproducible,
+    // so the test only verifies that the result was computed successfully and that result has the correct JSON format.
+    serde_json::from_str::<IdentityRecoveryRequestResult>(&res_json).unwrap();
```
Also here.

bisgardo replied:
done

abizjak commented on tests/lib_test.rs:
```diff
@@ -165,8 +191,381 @@ fn testnet_verifiable_credential_signing_key() {
+
+    let res = account_credential_deployment_hash_hex(input, 9_999_999_999).unwrap();
+    assert_eq!(
+        res,
+        "f5448faedc5011d1d6a622db7b6790d3ae1a06a331675717f62b66f9e33d1056".to_string(),
```
```suggestion
        res.as_str(),
        "f5448faedc5011d1d6a622db7b6790d3ae1a06a331675717f62b66f9e33d1056",
```

bisgardo replied:
done

abizjak commented on tests/lib_test.rs:
```diff
@@ -165,8 +191,381 @@ fn testnet_verifiable_credential_signing_key() {
+
+#[test]
+fn account_credential_deployment_concordium_ip() {
+    let res = account_credential(AccountCredentialParameters {
```
I think these tests are not maintainable. It is unclear how the values for them were produced, and these massive hex strings are difficult to navigate.

We have data structures for generating all of these values, including JSON serialization of them. For example we have globalcontex generation, IpInfo is a data structure with JSON serialization, etc.

So would it not be substantially easier to maintain compatibility with everything if those data structures were used to construct these values and then if you really insist on using this UniFFI then, in line with what we do in other places, we'd have TryFrom instances to convert between the imprecise UniFFI generated types and the precise concordium-base types that ultimately need to be constructed.

This would also then statically check that changes in base are correctly caught

bisgardo replied:
The blobs were created with the example wallet in the Java SDK because I didn't know the proper types initially. I agree that they're hideous and would like to do a followup to generate them properly instead. But for now I need to proceed with the SDK :)

abizjak commented on src/lib.rs:
```diff
@@ -137,19 +151,382 @@ pub fn get_verifiable_credential_public_key(
 }
+
+// TODO: Use 'derivative' to implement Debug in a way that doesn't include fields containing secrets
+//       and log the serialized debug string into the error message.
```
What is to become of this?

bisgardo replied:
Will address this in a follow-up.

bisgardo committed:
Add docstrings to Rust and UDL types and functions

In the Rust code, the docstrings explain the link between the types generated from UDL and the "original" ones from the core Concordium libraries that they bridge to. To properly refer to types located in concordium-base in the docstrings, an explicit dependency to that library has been added.

The docstrings on the UDL types are mostly what I could find in the docstrings on the original types.

As the dependency to concordium-base was added, the struct `Versioned` in `lib_test` was replaced by its original.

abizjak reviewed (approved):
Discussed offline, in the interest of moving things along we'll go with the current solution of manually rust defined types + JSON conversion.

bisgardo committed:
Test: Validate constructed identity requests

Before this we only verified that the result has the correct JSON format.
